### ISSUE TITLE
Task 4 apply external force on sphere

### ISSUE THREAD

**Issue by vullnetu** [cmm-21/a5#18]
I have implemented task 4 and it seems to work fine. However, If I start to drag the sphere (apply external force to it) it jumps away from my screen. I noticed that it has something to do with the update in the angular velocity (w'=w'+I^-1*(ra x J). Is my implementation wrong or is our model not laid out  for this
see below the video
[video](https://youtu.be/nzmLrDThjlo)

**Comment by eastskykang** [comment]
It could be possible that you are giving wrong value for **r**<sub>a</sub> (as well as d**r**<sub>a</sub>/dt)

Remember, the bottom of the ball is not a fixed point. It changes as the ball rotates.

**Comment by eastskykang** [comment]
This is what I got from my implementation: https://youtu.be/6RTHyJI2zUc

Some interpretations: 
- when the epsilon = 0, after the collision, the velocity of the contact point should be zero, thus the ball will start roll over the ground.

**Comment by vullnetu** [comment]
> It could be possible that you are giving wrong value for **r**a (as well as d**r**a/dt)
> 
> Remember, the bottom of the ball is not a fixed point. It changes as the ball rotates.

I agree the bottom of the ball is changing but **r**a is expressed in world coordinates and it is defined from the COM of the rigid body to the contact point. Therefore, I see  **r**a to be constant.

**Comment by eastskykang** [comment]
@vullnetu Yes. That's correct. Then now the question is how to get d**r**<sub>a</sub>/dt by using code base we have.

**Comment by iqiac** [comment]
Is this necessary to get full points or is it sufficient if we can recreate the demo?
I mean I implemented everything, but didn't change the lowest point as the sphere turns.

**Comment by vullnetu** [comment]
just to clarify with "pose" in the hints you mean position and orientation of the rigid body ?

**Comment by eastskykang** [comment]
@vullnetu yes, you're right. 

@iqiac okay, let's do this, I admit my instruction was a bit vague. I will give you full point if your implementation reproduce exactly same result with the demo I posted. You don't need to worry about external force. 

But, please try to modify your code to make it work with external force (and torque) and angular velocity > 0 if you have some time (and motivation). It does not require much effort.

**Comment by vullnetu** [comment]
For anyone wanting to solve this. Look up the definition of Urel closely, especially its direction. The extra [documents](https://www.cs.cmu.edu/~baraff/pbm/rigid2.pdf) on the readme file helped me to understand my mistake.

**Comment by iqiac** [comment]
@eastskykang Thanks, I'll definitely try later. It's just right now I'm pretty occupied with other subjects.
